fix #6, add transaction api

manuzhang opened the pull request.

clockfly commented:
Good work, it will require some time to review it.

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/user/example/processor/WindowAverageProcessor.scala:
```diff
@@ -0,0 +1,57 @@
+  extends PersistentTask[AveragedValue](taskContext, conf) {
+
+  private val windowSize = conf.getLong(PersistentStateConfig.STATE_WINDOW_SIZE)
+    .getOrElse(throw new RuntimeException("state window size not configured"))
```
We should never throw RuntimeException in actor.

Exception should also be avoided.

clockfly replied:
You can change the default constructor to make the dependencies explicit.

like this:
```
class WindowAverageProcessor(windowConfig: WindowConfig, taskContext : TaskContext, conf: UserConfig) {
  def this (taskContext: TaskContext, conf: UserConfig) = {
      this(conf.getLong(PersistentStateConfig.STATE_WINDOW_SIZE), context, conf)
  }
+  extends PersistentTask[AveragedValue](taskContext, conf) {

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/user/example/processor/WindowAverageProcessor.scala:
```diff
@@ -0,0 +1,57 @@
+class WindowAverageProcessor(taskContext : TaskContext, conf: UserConfig)
+  extends PersistentTask[AveragedValue](taskContext, conf) {
+
+  private val windowSize = conf.getLong(PersistentStateConfig.STATE_WINDOW_SIZE)
```
Maybe wrap windowSize and windowStep  in another objectt is better?

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/user/example/processor/WindowAverageProcessor.scala:
```diff
@@ -0,0 +1,57 @@
+    state.update(message.timestamp, value)
+  }
+
+  private[example] def getState: Option[AveragedValue] = {
```
If getState is only used for UT, I suggest removing it.

val state is no much difference with  def getState

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/user/example/WindowAverage.scala:
```diff
@@ -0,0 +1,68 @@
+import org.apache.gearpump.util.Graph
+import org.apache.gearpump.util.Graph.Node
+
+object WindowAverage extends App with ArgumentsParser {
```
How about making the name more explicit, like WindowAverageApp, so that people will not confuse it with internal impl class related with window logic.

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/user/api/PersistentTask.scala:
```diff
@@ -0,0 +1,56 @@
+   *  this is for user to get PersistentState
+   *  user should define or import an implicit algebird monoid
+   */
+  protected def persistentState(implicit monoid: Monoid[T]): MonoidState[T] =
```
I think the return type should be PersistentState, instead of MonidState.

As long as user implement PersistentState's interface, user should be allowed to define its own PersistentState task.

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/user/api/PersistentTask.scala:
```diff
@@ -0,0 +1,56 @@
+   * this is for user to get PersistentState with window functions
+   * user should define or import an implicit algebird group
+   */
+  protected def persistentState(windowSize: Long, windowStep: Long)(implicit group: Group[T]): WindowState[T] =
```
I don't know whether it is a good idea to expose 
```persistentState(windowSize: Long, windowStep: Long)```
? Why not using 
```
def persistentState(implicit monoid: Monoid[T])
```
?

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/user/api/PersistentTask.scala:
```diff
@@ -0,0 +1,56 @@
+   *  user should define or import an implicit algebird monoid
+   */
+  protected def persistentState(implicit monoid: Monoid[T]): MonoidState[T] =
+    new DefaultState(monoid, checkpointManager, taskContext)
```
I don't like the idea to define some default state here.
I think it is totally upto the developer about what state to use, the framework should make no assumption on this.

User can define its own persistentState or use framework provided persistentState.

And, user should also be able to combine multiple persstentState together to be a single persistentState.

Consider a use case, where use want to have three persistent states, 
1. a monid state of long
2. a monid state of CMS
3. a window state

How current design are going to support this? 

An alternative solution will be allow user to extend PersistentState, like this:
```
class MyState extends PersistentState {

private val longMonidState
private val cmsMonidState
private val windowState 

 ... 
other persistentState interface
}

```

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/system/impl/WindowState.scala:
```diff
@@ -0,0 +1,145 @@
+   *    upperBound1 = step * Nmin1 > t
+   *    upperBound2 = step * Nmin2 + size > t
+   */
+  private[impl] def getInterval(timestamp: TimeStamp, checkpointTime: TimeStamp): Interval = {
```
Please make this function clearer.

what kind interval? purpose?

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/system/impl/WindowManager.scala:
```diff
@@ -0,0 +1,46 @@
+
+import org.apache.gearpump.TimeStamp
+
+class WindowManager(val windowSize: Long, val windowStep: Long) {
```
PLease explain what this class is used for?

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/system/impl/DefaultState.scala:
```diff
@@ -0,0 +1,53 @@
+  val LOG: Logger = LogUtil.getLogger(classOf[DefaultState[_]])
+}
+
+class DefaultState[T](monoid: Monoid[T],
```
Please make a better name, so that it is more easy to understand the difference with WindowState.

manuzhang commented:
thanks @clockfly. I've updated the pr addressing your comments and also modifying PersistentState and PersistentTask APIs.

I've been thinking really hard on this design and actually going back and forth several times for the past month in previous versions of this pr. 

Finally, this is what I feel most comfortable.

`PersistentTask` provides users with transaction support by the framework. If users have their own ways to transaction they are free to write their own transactional tasks. To use `PersistentTask`, users should pass in a `PersistentState`, which could either be the framework's built-in `UnboundedState` and `WindowState` or their own implementations.
`PersistentTask` hides `onStart` and `onNext` (where the framework recovers and checkpoints state in the background) from users but provides extra hooks for users to invoke `PersistentState` API. Currently, we have `processMessage` which is called in `onNext` and we could also add one for `onStart`. 

As to support combining states, I prefer your second solution. The framework only knows one `Persistent[T]` although that `T` could combine multiple user states. Only binary data will be transferred between user states and the framework. I think this keeps our design simple.

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/system/impl/UnboundedState.scala:
```diff
@@ -0,0 +1,53 @@
+/**
+ * this is a MonoidState with no boundary
+ */
+class UnboundedState[T](monoid: Monoid[T]) extends MonoidState[T](monoid) {
```
I prefer the name MonidState than UnboundedState

clockfly replied:
Addition.

I don't like the word boundary, as I don't know what boundary means here, time boundary? I know you want to differentiate from WindowState, maybe NonWindowState better?

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/system/impl/WindowManager.scala:
```diff
@@ -0,0 +1,54 @@
+ * this is used in window applications
+ * it keeps the current window and slide ahead when the window expires
+ */
+class WindowManager(val windowSize: Long, val windowStep: Long) {
```
Maybe name Window is better instead of WindowManager.

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/system/impl/WindowManager.scala:
```diff
@@ -0,0 +1,54 @@
+    clock >= (startTime + windowSize)
+  }
+
+  def getCurrentWindow: (TimeStamp, TimeStamp) = {
```
Maybe windowRange is better than getCurrentWindow

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/system/impl/CheckpointManager.scala:
```diff
@@ -0,0 +1,56 @@
+import org.apache.gearpump.streaming.state.system.api.CheckpointStore
+
+
+class CheckpointManager(checkpointInterval: Long,
```
Maybe name CheckpointWorker or CheckpointHandler is better?

As this don't manage checkpoints.

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/system/impl/WindowState.scala:
```diff
@@ -0,0 +1,163 @@
+ */
+class WindowState[T](group: Group[T],
+                     taskContext: TaskContext,
+                     windowManager: WindowManager)
```
It will make sense if rename windowManager to window.

Then the window concept perfectly match WindowState.

clockfly commented on experiments/state/src/main/scala/org/apache/gearpump/streaming/state/user/example/processor/GenNumProcessor.scala:
```diff
@@ -0,0 +1,41 @@
+import org.apache.gearpump.cluster.UserConfig
+import org.apache.gearpump.streaming.task.{StartTime, Task, TaskContext}
+
+class GenNumProcessor(taskContext : TaskContext, conf: UserConfig) extends Task(taskContext, conf) {
```
Maybe name NumberGeneratorProcessor better?

clockfly commented:
Good work!

Only added a few small suggestions.

+1

manuzhang commented:
thanks @clockfly. updated against your comments.
